libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -1,79 +1,141 @@
-    private val tripSession: TripSession,
-    threadController: ThreadController,
+    private val currentLegIndexProvider: () -> Int,
     private val routeDiffProvider: DirectionsRouteDiffProvider = DirectionsRouteDiffProvider(),
+    private val localDateProvider: () -> Date
 ) {
 
     internal companion object {
         internal const val LOG_CATEGORY = "RouteRefreshController"
     }
 
-    private val routerRefreshTimer = MapboxTimer(threadController).apply {
-        restartAfterMillis = routeRefreshOptions.intervalMillis
+    suspend fun refresh(routes: List<NavigationRoute>): List<NavigationRoute> {
+        return if (routes.firstOrNull()?.routeOptions?.enableRefresh() == true) {
+            val result = routes.toMutableList()
+            val routeToRefresh = result.first()
+            result[0] = tryRefreshingUntilRouteChanges(routeToRefresh)
+            result
+        } else waitForever("refresh isn't enabled")
     }
 
-    private var currentRequestId: Long? = null
-
-    /**
-     * If route refresh is enabled, attach a refresh poller.
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun restart(route: NavigationRoute, onRoutesRefreshed: (List<NavigationRoute>) -> Unit) {
-        stop()
-        if (route.routeOptions.enableRefresh() == true) {
-            routerRefreshTimer.startTimer {
-                refreshRoute(route, onRoutesRefreshed)
+    private suspend fun tryRefreshingUntilRouteChanges(
+        initialRoute: NavigationRoute
+    ): NavigationRoute {
+        while (true) {
+            val refreshed = refreshRoute(initialRoute)
+            if (refreshed != initialRoute) {
+                return refreshed
             }
         }
     }
 
-    /**
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun stop() {
-        routerRefreshTimer.stopJobs()
-        currentRequestId?.let {
-            directionsSession.cancelRouteRefreshRequest(it)
-            currentRequestId = null
+    private suspend fun refreshRoute(
+        route: NavigationRoute
+    ): NavigationRoute = coroutineScope {
+        val routeLegs = route.directionsRoute.legs()
+        require(routeLegs != null) { "Can't refresh route without legs" }
+
+        var timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+        for (attempts in 0..2) {
+            timeUntilNextAttempt.await()
+            timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+            val refreshedRoute = withTimeoutOrNull(routeRefreshOptions.intervalMillis) {
+                refreshRouteOrNull(route)
+            }
+            if (refreshedRoute != null) {
+                timeUntilNextAttempt.cancel()
+                return@coroutineScope refreshedRoute
+            }
         }
+        timeUntilNextAttempt.cancel()
+        removeTrafficAnnotations(route, routeLegs)
     }
 
-    private fun refreshRoute(
+    private fun removeTrafficAnnotations(
```

Comment:
```suggestion
    private fun clearOutdatedAnnotationsAndIncidents(
```
nit, or even separate into two functions.

Reply:
Replaced by `removeExpiringDataFromRoute`